From 4a0eef75eddbe10bebe1f723095c449a2ce4b552 Mon Sep 17 00:00:00 2001
From: Cristina Ciocan <cristina-mihaela.ciocan@nxp.com>
Date: Wed, 5 Jul 2017 20:26:39 +0300
Subject: [PATCH 2238/5242] MLK-15986-8: video: fbdev: Add 64bit framebuffer
 support for ADV7535

commit  03e5394a4acf6932eb00d000151efd08d3c926c5 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds support for ADV7535  MIPI-to-HDMI converter for 64bit
platforms.

ADV7535 driver changes are added from Fancy Fang's commit ca389b0895c9
("MLK-15322-11: video: fbdev: adv7535: enable adv7535 driver").

Signed-off-by: Cristina Ciocan <cristina-mihaela.ciocan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/Kconfig   |    2 +-
 drivers/video/fbdev/mxc/adv7535.c |   26 ++++++++++++++++++++++++++
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/drivers/video/fbdev/mxc/Kconfig b/drivers/video/fbdev/mxc/Kconfig
index f135081..1e67cf5 100644
--- a/drivers/video/fbdev/mxc/Kconfig
+++ b/drivers/video/fbdev/mxc/Kconfig
@@ -99,7 +99,7 @@ config FB_MXC_DCIC
 config FB_MXC_ADV7535
 	tristate "ADI ADV7535 support"
 	depends on I2C
-	depends on FB_MXC_MIPI_DSI
+	depends on FB_MXC_MIPI_DSI || FB_MXC_MIPI_DSI_NORTHWEST
 	help
 	  Driver support for the ADV7535 DSI-to-HDMI module
 
diff --git a/drivers/video/fbdev/mxc/adv7535.c b/drivers/video/fbdev/mxc/adv7535.c
index 3035dc5..c232810 100644
--- a/drivers/video/fbdev/mxc/adv7535.c
+++ b/drivers/video/fbdev/mxc/adv7535.c
@@ -22,6 +22,8 @@
 #define ADV7535_CHIP_REVISION		0x0
 #define ADV7535_DSI_CEC_ADDR		0xE1
 
+#define TEST_PATTERN_ENABLE    0
+
 struct adv7535_info {
 	int rev;
 	struct device *dev;
@@ -140,8 +142,18 @@ static int adv7535_vmode_cfg(struct adv7535_info *info)
 		       fb_vmode->lower_margin + fb_vmode->vsync_len;
 
 	client = info->i2c_dsi_cec;
+#ifdef CONFIG_FB_IMX64
+	adv7535_write_reg(0x1C, 0x40); /* 4 Data Lanes */
+#else
 	adv7535_write_reg(0x1C, 0x20); /* 2 Data Lanes */
+#endif
+
+#if TEST_PATTERN_ENABLE
+	adv7535_write_reg(0x55, 0x80);
+	adv7535_write_reg(0x16, 0x1C);
+#else
 	adv7535_write_reg(0x16, 0x00); /* Pixel Clock  */
+#endif
 	adv7535_write_reg(0x27, 0xCB); /* INT_TIMING_GEN */
 
 	/* video mode settings */
@@ -193,11 +205,21 @@ static int adv7535_vmode_cfg(struct adv7535_info *info)
 	client = info->i2c_main;
 	adv7535_write_reg(0xAF, 0x16); /* HDMI Output */
 	adv7535_write_reg(0x55, 0x10); /* AVI Info-frame */
+#ifdef CONFIG_FB_IMX64
+	adv7535_write_reg(0x56, 0x28); /* 16:9 */
+#else
 	adv7535_write_reg(0x56, 0x18);
+#endif
 	adv7535_write_reg(0x40, 0x80); /* GCP Enable */
 	adv7535_write_reg(0x4C, 0x04); /* 24bpp */
 	adv7535_write_reg(0x49, 0x00);
+
+#ifdef CONFIG_FB_IMX64
+	/* low refresh rate */
+	adv7535_write_reg(0x4A, 0x8C);
+#else
 	adv7535_write_reg(0x17, 0x60); /* VS & HS Low Polarity, DE disabled */
+#endif
 
 	/*TODO Audio Setup */
 
@@ -290,7 +312,9 @@ static int adv7535_probe(struct i2c_client *client,
 	if ((ret = adv7535_vmode_cfg(info)))
 		goto err3;
 
+#ifndef CONFIG_FB_IMX64
 	pm_runtime_enable(info->dev);
+#endif
 
 	dev_info(dev, "adv7535 probe finished\n");
 
@@ -312,7 +336,9 @@ static int adv7535_remove(struct i2c_client *client)
 	info = i2c_get_clientdata(client);
 	i2c_set_clientdata(client, NULL);
 
+#ifndef CONFIG_FB_IMX64
 	pm_runtime_disable(info->dev);
+#endif
 
 	i2c_unregister_device(info->i2c_dsi_cec);
 
-- 
1.7.9.5

